path-timestamp.sh

```
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+# Expose timestamp of given path
+#
+# Usage: add this to crontab:
+#
+# */5 * * * * prometheus path-timestamp.sh /var/lib/prometheus | sponge /var/lib/node_exporter/path-timestamp.prom
+#
+# Author: Antoine Beaupré <anarcat@debian.org>
+
+echo "# HELP node_path_modification_timestamp_seconds Last change timestamp"
+echo "# TYPE node_path_modification_timestamp_seconds gauge"
+for path in "$@"; do
+    printf 'node_path_modification_timestamp_seconds{path="%s"}' "$path"
+    stat -c %Y "$path"
```

```suggestion
    if ! stat -c %Y "$path"; then
      # clear line for next metric
      echo 0
    fi
```

If a file does not exist, this line can cause the output to become invalid like the following:

~~~
$ ./path-timestamp.sh /home/tor/dev/tpa/tor-puppet/hiera.yaml environment.conf path-timestamp.sh ext_fs_error_count.sh 
# HELP node_path_modification_timestamp_seconds Last change timestamp
# TYPE node_path_modification_timestamp_seconds gauge
node_path_modification_timestamp_seconds{path="/home/tor/dev/tpa/tor-puppet/hiera.yaml"} 1734046134
node_path_modification_timestamp_seconds{path="environment.conf"} 1717708435
node_path_modification_timestamp_seconds{path="path-timestamp.sh"} 1736455719
node_path_modification_timestamp_seconds{path="ext_fs_error_count.sh"} 1736452903
~~~

or without stderr we have:

~~~
$ ./path-timestamp.sh /home/tor/hiera.yaml environment.conf path-timestamp.sh ext_fs_error_count.sh 2>/dev/null
# HELP node_path_modification_timestamp_seconds Last change timestamp
# TYPE node_path_modification_timestamp_seconds gauge
node_path_modification_timestamp_seconds{path="/home/tor/hiera.yaml"} node_path_modification_timestamp_seconds{path="environment.conf"} 1717708435
node_path_modification_timestamp_seconds{path="path-timestamp.sh"} 1736455719
node_path_modification_timestamp_seconds{path="ext_fs_error_count.sh"} 1736452903

---

i implemented a different solution for this, where we set the timestamp to zero if stat(1) fails, which will allow us to alert on those cases as well.

---

great, so we had the same idea wrt setting it to 0, but your solution is more compact. and it's good that the comment above mentions why the value can be set to 0. :100:

---

oh. oops. sorry, i misread your suggested change and interpreted it as:

```
exit 0
```

not:

```
echo 0
```

as you wrote.

it seems like we're in violent agreement then! :)

what do you think of silencing the error from the stat failure? should we let those through? i was thinking maybe it would make diagnostics easier in case of failures, but then it might mess up the output if, for some reason stupid reason i can't take of, (e.g.) sponge(1) takes in stderr alongside stdout...